### ISSUE TITLE
Embedded DNS problem after renaming container

### ISSUE THREAD

**Issue by BediarSofiane** [docker/docker#22466]
Hello,
I am trying to launch two ubuntu containers and connect them with a bridge network. When I run  or create directly the two containers  with --name option then connect them to a network, it works pretty well. But when i do it step by step, i,e., Create ==>** rename** ==> connect ==> start , the containers don't know each other by name.
I'm using the version 1.10.
Thank you.

**Comment by thaJeztah** [comment]
Can you still reproduce this on docker 1.11? I tried to reproduce, but it seems to be working for me

```bash
# create a custom bridge network
docker network create testnet

# start a container and connect it to the network
docker run -d --net=testnet --name client nginx

# create, rename, connect, and start another container
docker create --name oldname nginx
docker rename oldname newname
docker network connect testnet newname
docker start newname


# try reaching the "renamed" container from inside the "client" container
docker exec -t client sh -c "ping -c3 newname"
PING newname (172.20.0.3): 56 data bytes
64 bytes from 172.20.0.3: icmp_seq=0 ttl=64 time=0.097 ms
64 bytes from 172.20.0.3: icmp_seq=1 ttl=64 time=0.070 ms
64 bytes from 172.20.0.3: icmp_seq=2 ttl=64 time=0.073 ms
--- newname ping statistics ---
3 packets transmitted, 3 packets received, 0% packet loss
round-trip min/avg/max/stddev = 0.070/0.080/0.097/0.000 ms


# and try reaching the "client" container from inside the "renamed" container
docker exec -t newname sh -c "ping -c3 client"
PING client (172.20.0.2): 56 data bytes
64 bytes from 172.20.0.2: icmp_seq=0 ttl=64 time=0.062 ms
64 bytes from 172.20.0.2: icmp_seq=1 ttl=64 time=0.068 ms
64 bytes from 172.20.0.2: icmp_seq=2 ttl=64 time=0.071 ms
--- client ping statistics ---
3 packets transmitted, 3 packets received, 0% packet loss
round-trip min/avg/max/stddev = 0.062/0.067/0.071/0.000 ms
```

**Comment by BediarSofiane** [comment]
Hi @thaJeztah , thank you for your answer.
In fact i didn't try with v1.11. I'll do it now. However my test was not exactly the same as yours. Here is what i exactly tested:

# create a custom bridge network
-> docker network create testnet

# create  the first container
-> docker create -it ubuntu /bin/bash
-> docker rename [the container's Id] ub1
->docker network connect testnet ub1

# create  the second container
-> docker create -it ubuntu /bin/bash
-> docker rename [the container's Id] ub2
->docker network connect testnet ub2

# start the two containers
-> docker start ub1
-> docker start ub2

# testing ping
-> docker attach ub1
-> ping ub2
(the ping doesn't work).

**Comment by BediarSofiane** [comment]
I've tested it on v1.11 and it still not working.

**Comment by sanimej** [comment]
@BediarSofiane Docker service discovery is enabled only for named containers. When you create it as anonymous container (without --name) and later rename it, its still treated as anonymous and hence its not discoverable. This can be fixed. But to avoid this issue you can pass --name parameter when creating the container.

**Comment by jmzwcn** [comment]
@sanimej ,  could I try to fix this issue? and let me know more about solution?

**Comment by sanimej** [comment]
@jmzwcn Yes, please go ahead.

For containers without --name the anonymous flag is set using this call in BuildCreateEndpointOptions()
````
        if n.Name() == defaultNetName || container.NetworkSettings.IsAnonymousEndpoint {
                createOptions = append(createOptions, libnetwork.CreateOptionAnonymous())
        }
````
daemon/rename.go  ContainerRename() handles the rename. You can add a libnetwork function to reset the flag and call it on a successful rename.

**Comment by jmzwcn** [comment]
#dibt

**Comment by jmzwcn** [comment]
#dibs

**Comment by BediarSofiane** [comment]
Thank you @sanimej,
I realised that. 
I'm wondering why does the rename work well when we rename a container that was created with --name (the name is updated in DNS, see example of @thaJeztah ) and doesn't work with a container that is created without --name   (except if it is not the same part of code that is  executed).

Another thing: since the container created without --name is attributed a generated name. I think that it will make more sense to treat this generated name exactly as we treat the custom name provided with --name, so that, even if we don't rename this container (created without --name) his name should  be known by the DNS server.

**Comment by thaJeztah** [comment]
@BediarSofiane that's a design choice; if you want a container to be discoverable, you need to give it a name. Other containers won't know the name anyway if it's randomly generated

**Comment by jmzwcn** [comment]
I created a PR and should be the smallest change I think. https://github.com/docker/libnetwork/pull/1154

Please let me know if there is any problem.

**Comment by BediarSofiane** [comment]
Okay thank you for your answer :+1:

**Comment by jmzwcn** [comment]
@sanimej, @thaJeztah, @BediarSofiane
I created two PRs for this issue, including daemon side and libnetwork side. #1164, #22716.

I have verified in local dev-build and seems not find issue any more.

Could you please review them and let me know if there is any problem?

**Comment by jmzwcn** [comment]
@sanimej, @thaJeztah
The both PRs are passed finally. Any comments for the change? Thanks!

**Comment by jmzwcn** [comment]
Two changes are both merged, I think this issue could be closed.

**Comment by thaJeztah** [comment]
@jmzwcn are the libnetwork changes already included in master? we may have to wait with closing until those are vendored in

**Comment by jmzwcn** [comment]
Yes, https://github.com/docker/libnetwork/pull/1164 have been merged into master.

**Comment by thaJeztah** [comment]
@jmzwcn it was merged to master in libnetwork, but docker does not yet have it vendored in (if I see if correctly) https://github.com/docker/docker/blob/1f9e3815aa4ac7eaa707f3a9b436f23c29355a8f/vendor/src/github.com/docker/libnetwork/endpoint.go#L536

**Comment by jmzwcn** [comment]
Well, thanks for your clarify.

**Comment by yongtang** [comment]
@thaJeztah @jmzwcn I was looking into updating libnetwork to bring in ipvs and load balancing with pull request #23266. I think it will help close this embedded dns problem once it is merged.

**Comment by sanimej** [comment]
Fix for  this is already in docker master. Closing the issue..
